docs/contributing/retries-and-waiters.md

```diff
@@ -0,0 +1,474 @@
+- IAM Role being reported as not having sufficient permissions (inline or attached role permissions)
+
+Each AWS service API (and sometimes even operations within the same API) varies in the implementation of these errors. To handle them, it is recommended to use the Operation Specific Error Retries pattern. The Terraform AWS Provider implements a standard timeout constant of two minutes in the `aws/internal/service/iam/waiter` package which should be used for all retry timeouts associated with IAM errors. This timeout was derived from years of Terraform operational experience with all AWS APIs.
+
+```go
+// aws/resource_example_thing.go
+
+import (
+	// ... other imports ...
+	// By convention, cross-service waiter imports are named {SERVICE}waiter
+	iamwaiter "github.com/terraform-providers/terraform-provider-aws/aws/internal/service/iam/waiter"
+)
+
+// ... Create and typically Update function ...
+	err := resource.Retry(iamwaiter.PropagationTimeout, func() *resource.RetryError {
+		_, err := conn./* ... AWS Go SDK operation with IAM eventual consistency errors ... */
+
+		// Example retryable condition
+		// This must be updated to match the AWS service API error code and message.
+		if tfawserr.ErrMessageContains(err, /* error code */, /* error message */) {
+			return resource.RetryableError(err)
+		}
+
+		if err != nil {
+			return resource.NonRetryableError(err)
+		}
+
+		return nil
+	})
+
+	if tfresource.TimedOut(err) {
+		_, err = conn./* ... AWS Go SDK operation with IAM eventual consistency errors ... */
+	}
+
+	if err != nil {
+		return fmt.Errorf("... error message context ... : %w", err)
+	}
+```
+
+### Resource Lifecycle Retries
+
+When the eventual consistency lies with the existence or configuration state of the remote system component after an operation, this is referred to as resource lifecycle eventual consistency. A common example of this includes creating a resource and immediately reading it (or attributes of it) where those subsequent operations return a "not found" error.
+
+When encountering this type of situation, every operation after the initial operation must account for the condition, which can be tricky depending on the resource logic itself. Most resources should be written where the `Create` function will return the `Read` function to fill in all computed attributes and ensure the configuration was applied correctly. In these cases, the `Read` function will need additional logic to overcome the temporary condition on resource creation. For eventually consistent resources, these "not found" type errors can still occur in the `Read` function even after implementing [Resource Lifecycle Waiters](#resource-lifecycle-waiters) in the Create function for the resource itself.
+
+```go
+// aws/internal/service/{SERVICE}/waiter/consts.go (created if does not exist)
+
+package {SERVICE}
+
+const (
+	// Maximum amount of time to wait for Example eventual consistency on creation
+	ExampleCreationTimeout = 2 * time.Minute
```

Comment:
If _Example_ is the service name and we are in a service-specific package, should this be `CreationTimeout = 2 * time.Minute` instead?

Reply:
Fixed in latest commit.